python/cudf/cudf/_lib/scalar.pyx

```diff
@@ -93,41 +120,73 @@ cdef class Scalar:
                 f"{type(value).__name__} to cudf scalar"
             )
 
+    def _get_device_value(self):
+        if pd.api.types.is_string_dtype(self.dtype):
+            result = _get_py_string_from_string(self.c_value)
+        elif pd.api.types.is_numeric_dtype(self.dtype):
+            result = _get_np_scalar_from_numeric(self.c_value)
+        elif pd.api.types.is_datetime64_dtype(self.dtype):
+            result = _get_np_scalar_from_timestamp64(self.c_value)
+        elif pd.api.types.is_timedelta64_dtype(self.dtype):
+            result = _get_np_scalar_from_timedelta64(self.c_value)
+        else:
+            raise ValueError(
+                "Could not convert cudf::scalar to a Python value"
+            )
+        return result
+
     @property
     def dtype(self):
         """
         The NumPy dtype corresponding to the data type of the underlying
         device scalar.
         """
-        cdef libcudf_types.data_type cdtype = self.c_value.get()[0].type()
+        cdef libcudf_types.data_type cdtype = self.get_raw_ptr()[0].type()
         return cudf_to_np_types[<underlying_type_t_type_id>(cdtype.id())]
 
     @property
     def value(self):
         """
         Returns a host copy of the underlying device scalar.
         """
-        if pd.api.types.is_string_dtype(self.dtype):
-            return _get_py_string_from_string(self.c_value)
-        elif pd.api.types.is_numeric_dtype(self.dtype):
-            return _get_np_scalar_from_numeric(self.c_value)
-        elif pd.api.types.is_datetime64_dtype(self.dtype):
-            return _get_np_scalar_from_timestamp64(self.c_value)
-        elif pd.api.types.is_timedelta64_dtype(self.dtype):
-            return _get_np_scalar_from_timedelta64(self.c_value)
+        if self._is_host_value_current():
+            return self._host_value
         else:
-            raise ValueError(
-                "Could not convert cudf::scalar to a Python value"
-            )
+            result = self._get_device_value()
+            self._host_value = result
+            return result
+
+    def _sync(self):
```

Comment:
What is the general behavior when neither the host or device has an initialized value?

Reply:
Basically, there shouldn't be any way of actually creating that object via normal means. We can either make a scalar from the host, in which case the host value gets set, or the device in which case the device value gets set. If you were to make one from the host and then deliberately invalidate it, maybe by instantiating it and then setting `_host_value` to be none - you'd probably get a segfault, because in lieu of a valid host value the scalar will assume the device value has to be current and try and copy it to host at that point. 

I'll think about how to deal with this, because as I think you have said in the past, if it can happen we have to assume it will...